### PR TITLE
Ecs exec cmd integ test

### PR DESCRIPTION
**Pull request by angelcar**
### Summary

Add integration test for exec command agent.

### Implementation details

A new, scratch-based docker image was added for this integ test in order to be consistent with design doc which states exec command agent should be compatible with scratch-based images. The image has a pre-baked `sleep` command built using golang, as part of the docker build process.

The actual integ test only validates that the expected bindmounts are present in the container. This is done by invoking container inspect docker API and verifying that the expected bind mounts are in the response. This means that ECS Agent has added the required bind mounts for exec command agent to run.

For now I am skipping the test until the actual exec command binaries are present in the host.

note: Some minor changes are also included in this PR, such as removal of unused arguments.

### Testing

New tests cover the changes: yes

Ran `make run-integ-tests` and manually verified that the container was created and had the expected bind mounts. 
After that, I added the t.Skip statement to the test.

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/api/task/task_linux.go [hunk 1]
```diff
@@ -272,7 +273,7 @@ func (task *Task) initializeExecCommandAgentResources(cfg *config.Config) error
 		return err
 	}
 
-	execCommandAgentBinNames := []string{execCommandAgentBinName, execCommandAgentSessionWorkerBinName, execCommandAgentSessionLoggerBinName}
+	execCommandAgentBinNames := []string{ExecCommandAgentBinName, ExecCommandAgentSessionWorkerBinName, ExecCommandAgentSessionLoggerBinName}
```
Comment (fierlion) [inline agent/api/task/task_linux.go:276]: The SessionLogger is not necessary for Linux.

File: agent/api/task/task_linux.go [hunk 2]
```diff
@@ -53,16 +53,16 @@ const (
...
 
 	// TODO: [ecs-exec] decide if this needs to be configurable or put in a specific place in our optimized AMIs
-	execCommandAgentHostBinDir           = "/home/ec2-user/ssm-agent/linux_amd64"
-	execCommandAgentBinName              = "amazon-ssm-agent"
-	execCommandAgentSessionWorkerBinName = "ssm-session-worker"
-	execCommandAgentSessionLoggerBinName = "ssm-session-logger"
+	ExecCommandAgentHostBinDir           = "/home/ec2-user/ssm-agent/linux_amd64"
+	ExecCommandAgentBinName              = "amazon-ssm-agent"
+	ExecCommandAgentSessionWorkerBinName = "ssm-session-worker"
+	ExecCommandAgentSessionLoggerBinName = "ssm-session-logger"
```
Comment (fierlion) [inline agent/api/task/task_linux.go:59]: The session-logger is not necessary for linux.

File: agent/engine/engine_unix_integ_test.go [hunk 3]
```diff
@@ -1526,3 +1530,111 @@ func TestFluentdTag(t *testing.T) {
...
+			dest:     filepath.Join(apitask.ExecCommandAgentContainerBinDir, apitask.ExecCommandAgentBinName),
+			readOnly: true,
+		},
+		{
+			source:   filepath.Join(apitask.ExecCommandAgentHostBinDir, apitask.ExecCommandAgentSessionWorkerBinName),
+			dest:     filepath.Join(apitask.ExecCommandAgentContainerBinDir, apitask.ExecCommandAgentSessionWorkerBinName),
+			readOnly: true,
+		},
+		{
+			source:   filepath.Join(apitask.ExecCommandAgentHostBinDir, apitask.ExecCommandAgentSessionLoggerBinName),
```
Comment (fierlion) [inline agent/engine/engine_unix_integ_test.go:1591]: (ditto - no SessionLogger)

File: misc/exec-command-agent-test/linux.dockerfile [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+FROM golang:1.12 as build-env
+MAINTAINER Amazon Web Services, Inc.
+
+WORKDIR /go/src/sleep
+ADD ./sleep /go/src/sleep
+RUN go get -d -v ./...
+RUN CGO_ENABLED=0 GOOS=linux GOARCH=amd64 go build -installsuffix cgo -a -o /go/bin/sleep .
```
Comment (fierlion) [inline misc/exec-command-agent-test/linux.dockerfile:19]: cute static sleeper!

File: agent/api/task/task_linux_test.go [hunk 5]
```diff
@@ -1284,22 +1284,22 @@ func TestPostUnmarshalTaskWithExecCommandAgentEnabled(t *testing.T) {
 		assert.Equal(t, test.expectedVolumes, len(taskVolumes), "Should have created 6 task volumes")
 
 		// Check agent binary volumes
-		expectedBinaryNames := []string{execCommandAgentBinName, execCommandAgentSessionWorkerBinName, execCommandAgentSessionLoggerBinName}
+		expectedBinaryNames := []string{ExecCommandAgentBinName, ExecCommandAgentSessionWorkerBinName, ExecCommandAgentSessionLoggerBinName}
```
Comment (fierlion) [inline agent/api/task/task_linux_test.go:1287]: ditto no session-logger

**Review by fierlion — COMMENTED** [review]
nice overall -- 
main change is just to remove references to the `ssm-session-logger` (as it will presently only be required for Windows)

File: agent/api/task/task_linux.go [hunk 6]
```diff
@@ -262,7 +262,8 @@ func (task *Task) GetCredentialSpecResource() ([]taskresource.TaskResource, bool
 
 // initializeExecCommandAgentResources specifies the necessary volumes and mount points in all of the task containers in order for the
 // exec agent to run upon container start.
-func (task *Task) initializeExecCommandAgentResources(cfg *config.Config) error {
+// TODO: [ecs-exec] Should we validate the ssm agent binaries & certs are valid and fail here if they're not? (bind mount will succeed even if files don't exist in host)
```
Comment (sharanyad) [inline agent/api/task/task_linux.go:265]: feels like this should be part of agent capability advertisement.

File: agent/api/task/task_linux.go (hunk 1 above)
Comment (sharanyad) [reply agent/api/task/task_linux.go:276]: does the ExecAgent take care of logging?

File: misc/exec-command-agent-test/linux.dockerfile [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+FROM scratch
+MAINTAINER Amazon Web Services, Inc.
+COPY --from=build-env /go/bin/sleep /
```
Comment (sharanyad) [inline misc/exec-command-agent-test/linux.dockerfile:23]: nit: newline

**Review by sharanyad — COMMENTED** [review]
are we planning to add more validations to the test later other than just checking for mount points?

File: misc/exec-command-agent-test/linux.dockerfile (hunk 4 above)
Comment (angelcar) [reply misc/exec-command-agent-test/linux.dockerfile:19]: I wonder where I stole it from? :p

File: agent/api/task/task_linux.go (hunk 1 above)
Comment (angelcar) [reply agent/api/task/task_linux.go]: removed

File: agent/api/task/task_linux.go (hunk 2 above)
Comment (angelcar) [reply agent/api/task/task_linux.go]: removed

File: agent/engine/engine_unix_integ_test.go (hunk 3 above)
Comment (angelcar) [reply agent/engine/engine_unix_integ_test.go]: removed

File: agent/api/task/task_linux_test.go (hunk 5 above)
Comment (angelcar) [reply agent/api/task/task_linux_test.go]: removed

File: misc/exec-command-agent-test/linux.dockerfile (hunk 7 above)
Comment (angelcar) [reply misc/exec-command-agent-test/linux.dockerfile]: added

File: agent/api/task/task_linux.go (hunk 6 above)
Comment (angelcar) [reply agent/api/task/task_linux.go:264]: Is there any action item here? not exactly sure what this means :)

**Review by fierlion — APPROVED** [review]
changes look good.

File: agent/api/task/task_linux.go (hunk 6 above)
Comment (fierlion) [reply agent/api/task/task_linux.go:264]: we'll do some planning around this -- I think for now this is ok as a TODO.

**Review by sharanyad — APPROVED** [review]

**Pull request merged**